src/org/zalando/stups/friboo/user.clj

```diff
@@ -17,22 +18,76 @@
 
+; get team for service user
+(defcommand
+  fetch-service-teams
+  [service-user-url access-token user-id]
+  (let [team (-> (http/get (r/conpath service-user-url "/services/" user-id)
+                           {:oauth-token access-token
+                            :as :json})
+                 (get-in [:body :owner]))]
+    (vector {:name team}))) ; for consistency with human teams api
 
 (def get-teams
   "Cache team information for 5 minutes"
-  (memo/fifo fetch-teams (cache/ttl-cache-factory {} :ttl 300000) :fifo/threshold 100))
+  (memo/fifo fetch-teams
+             (cache/ttl-cache-factory {} :ttl 300000)
+             :fifo/threshold 100))
+
+(def get-service-teams
+  "Cache team info for service users 5 minutes"
+  (memo/fifo fetch-service-teams
+             (cache/ttl-cache-factory {} :ttl 300000)
+             :fifo/threshold 100))
+
+(defn trim-realm
+  "Trims leading and trailing slashes from the given realm.
+  If input is the top-level realm \"/\", leaves it as is."
+  [string]
+  (when string
+    (if (= string "/")
+      string
+      (clojure.string/replace string #"^\/*(.*?)\/*$" "$1"))))
+
+(defn require-realms
+  "Throws an exception if user is not in the given realms, else returns the user's realm"
+  [realms {:keys [tokeninfo]}]
+  (let [realm (trim-realm (get tokeninfo "realm"))
+        user-id (get tokeninfo "uid")]
+    (if (contains? realms realm)
+      realm
+      (do
+        (log/warn "ACCESS DENIED (unauthorized) because user is not in realms %s." realms)
+        (api/throw-error 403 (str "user not in realms '" realms "'")
+                         {:user            user-id
+                          :required-realms realms
+                          :user-realm      realm})))))
 
 (defn require-teams
   "Returns a set of teams, a user is part of or throws an exception if user is in no team."
   ([request]
-   (require-teams (:tokeninfo request) (require-config (:configuration request) :team-service-url)))
-  ([tokeninfo team-service-url]
-   (require-teams (get tokeninfo "uid") (get tokeninfo "access_token") team-service-url))
-  ([user-id token team-service-url]
+   (require-teams (:tokeninfo request)
+                  (require-config (:configuration request)
+                                  :team-service-url)
+                  (require-config (:configuration request)
+                                  :service-user-url)))
+  ([tokeninfo team-service-url service-user-url]
+   (require-teams (get tokeninfo "uid")
+                  (get tokeninfo "access_token")
+                  team-service-url
+                  service-user-url))
+  ([user-id token team-service-url service-user-url]
    (when-not user-id
      (log/warn "ACCESS DENIED (unauthenticated) because token does not contain user information.")
      (api/throw-error 403 "no user information available"))
-   (let [teams (get-teams team-service-url token user-id)]
+   (let [realm (require-realms #{"employees" "services"} token)
```

Comment:
Not sure if we should really require realms here, but if it's not one of those two, we cannot look up anything.